Common/include/toolboxes/CQuasiNewtonDriver.hpp

```diff
@@ -0,0 +1,253 @@
+ * via the normal equations as that is easy to make parallel over MPI, it
+ * may however be unstable with large sample sizes (compared to QR decomp).
+ * Usage: Allocate, store the initial solution (operator (i,j), default is 0),
+ * run the FP, store its result ("FPresult"), compute new solution, use it
+ * as the new input of the FP, run the FP, etc.
+ */
+template<class Scalar_t>
+class CQuasiNewtonDriver {
+public:
+  using Scalar = Scalar_t;
+  using Index = typename su2matrix<Scalar>::Index;
+  static_assert(std::is_floating_point<Scalar>::value,"");
+
+private:
+  using MPI_Wrapper = typename SelectMPIWrapper<Scalar>::W;
+
+  enum: size_t {BLOCK_SIZE = 1024};     /*!< \brief Loop tiling parameter. */
+  std::vector<su2matrix<Scalar> > X, R; /*!< \brief Input and residual history of the FP. */
+  su2matrix<Scalar> corr;               /*!< \brief Correction to the natural FP result. */
+  su2vector<Scalar> mat, rhs, sol;      /*!< \brief Matrix, rhs, and solution of the normal equations. */
+  Index iSample = 0;                    /*!< \brief Current sample index. */
+  Index nPtDomain = 0;                  /*!< \brief Local size of the history, considered in dot products. */
+
+  void shiftHistoryLeft() {
+    for (Index i = 1; i < X.size(); ++i) {
+      /*--- Swap instead of moving to re-use the memory of the first sample.
+       * This is why X and R are not stored as contiguous blocks of mem. ---*/
+      std::swap(X[i-1], X[i]);
+      std::swap(R[i-1], R[i]);
+    }
+  }
+
+  void computeNormalEquations() {
+    /*--- Size for the dot products. ---*/
+    const auto end = std::min<Index>(nPtDomain,corr.rows())*corr.cols();
+
+    mat = Scalar(0); rhs = Scalar(0);
+
+    /*--- Tiled part of the loop. ---*/
+    Index begin = 0;
+    while (end-begin >= BLOCK_SIZE) {
+      computeNormalEquations<BLOCK_SIZE>(mat, rhs, begin);
+      begin += BLOCK_SIZE;
+    }
+
+    /*--- Remainder of the loop. ---*/
+    if (begin != end) {
+      computeNormalEquations<0>(mat, rhs, begin, end-begin);
+    }
+
+    /*--- MPI reduction of the dot products. ---*/
+    if (nPtDomain < corr.rows()) {
+      const auto type = (sizeof(Scalar) < sizeof(double))? MPI_FLOAT : MPI_DOUBLE;
+
+      su2vector<Scalar> tmp(mat.size());
+      MPI_Wrapper::Allreduce(mat.data(), tmp.data(), iSample*(iSample+1)/2,
+                             type, MPI_SUM, MPI_COMM_WORLD);
+      mat = std::move(tmp);
+
+      MPI_Wrapper::Allreduce(rhs.data(), sol.data(), iSample,
+                             type, MPI_SUM, MPI_COMM_WORLD);
+      std::swap(rhs, sol);
+    }
+  }
+
+  template<Index StaticSize>
+  void computeNormalEquations(su2vector<Scalar>& mat,
+                              su2vector<Scalar>& vec,
+                              Index start,
+                              Index dynSize = 0) const {
+    /*--- Select either the static or dynamic size, optimizes inner loop. ---*/
+    const auto blkSize = StaticSize? StaticSize : dynSize;
```

Comment:
Because we tiled the loop, we have many work chunks with a static size, and one remainder chunk with dynamic size (not known during compilation).
This method is templated on the static size, so that we can "generate" two versions, for the version where the compiler knows the size of the loop it can generate much more efficient code because:
The size is a multiple of the SIMD length, and likely also of the loop unrolling count, which means only the necessary instructions are generated (better use of instruction cache), and size checks are not needed (less branching).